build_runner_core/lib/src/generate/single_step_reader_writer.dart

```diff
@@ -444,14 +444,41 @@ class SingleStepReaderWriter extends AssetReader
       if (nodePhase >= phase) {
         return PhasedValue.unavailable(before: '', untilAfterPhase: nodePhase);
       } else {
+        // If needed, trigger a build at an earlier phase.
+        if (node.generatedNodeState!.pendingBuildAction !=
+            PendingBuildAction.none) {
+          await _runningBuild.nodeBuilder(id);
+          node = _runningBuild.assetGraph.get(id)!;
+        }
         return PhasedValue.generated(
           atPhase: phase,
           before: '',
-          await _delegate.readAsString(id),
+          node.generatedNodeState!.isSuccessfulFreshOutput
+              ? await _delegate.readAsString(id)
+              : '',
         );
       }
     }
 
-    return PhasedValue.fixed(await _delegate.readAsString(id));
+    return PhasedValue.fixed(
+      await _delegate.canRead(id) ? await _delegate.readAsString(id) : '',
+    );
+  }
+
+  @override
+  bool hasChanged(AssetId id, {required int comparedToPhase}) {
+    if (comparedToPhase == phase) return false;
+    if (_runningBuild == null) return false;
+
+    final node = _runningBuild.assetGraph.get(id);
+    if (node == null) return false;
```

Comment:
Previously in this CL I saw that we deleted a file if reading it returned null or something.
Why does a null node here indicate a non-change?

Reply:
Null here means it's not in the asset graph, which means it's not a generated file; if you read it once, you'll get the same value by reading it again. So it only needs syncing once.

The dartdoc for the method now says Returns `false` for all other types of asset, including unknown assets." :)

Thanks.